gadget-container/gadgettracermanager/controller.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"os"
+
+	log "github.com/sirupsen/logrus"
+
+	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
+	// to ensure that exec-entrypoint and run can make use of them.
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+
+	gadgetkinvolkiov1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/controllers"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/biolatency"
+	networkpolicyadvisor "github.com/kinvolk/inspektor-gadget/pkg/gadgets/networkpolicy"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager"
+	//+kubebuilder:scaffold:imports
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+
+	utilruntime.Must(gadgetkinvolkiov1alpha1.AddToScheme(scheme))
+	//+kubebuilder:scaffold:scheme
+}
+
+func startController(node string, tracerManager *gadgettracermanager.GadgetTracerManager) {
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "0", // TCP port can be set to "0" to disable the metrics serving
+	})
+	if err != nil {
+		log.Errorf("unable to start manager: %s", err)
+		os.Exit(1)
+	}
+
+	gadgetRegistry := make(map[string]gadgets.Gadget)
+	gadgetRegistry["biolatency"] = &biolatency.Gadget{}
+	gadgetRegistry["network-policy-advisor"] = &networkpolicyadvisor.Gadget{}
```

Comment:
Before I knew about Kubernetes Finalizers, it was necessary for each gadget implementation to keep track of each trace because on trace deletion, the controller could only give the namespace/name of the resource. So I initially though of each `Gadget{}` as a singleton.

But with Kubernetes Finalizers, this is no longer the case, so it could be designed differently.

But I am not sure it should be done on the `start` operation: which code would create the new `&Gadget{}`?

Reply:
mmm, yes, it's not that clear. Maybe we can just create the gadget when the resource is created?

Reply:
Updated. It's no longer a singleton but using factories.